### PR TITLE
Fix: change to other modes from multi mode

### PR DESCRIPTION
**Pull request by hongduc-phan**
Fix: Switch to other selectionMode like SINGLE or NONE from MULTI.

#831

### CODE REVIEW

File: vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/AbstractGridMultiSelectionModel.java [hunk 1]
```diff
@@ -76,15 +76,15 @@ public AbstractGridMultiSelectionModel(Grid<T> grid) {
...
+        grid.getElement()
+                .insertChild(0, selectionColumn.getElement());
     }
 
     @Override
     protected void remove() {
         super.remove();
         deselectAll();
-        if (selectionColumn.getElement().getNode().isAttached()) {
+        if (selectionColumn.getElement() != null && getGrid().getElement().getChild(0).equals(selectionColumn.getElement())) {
```
Comment (tomivirkki) [inline vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/AbstractGridMultiSelectionModel.java]: Why is this null check needed now? The master version didn't seem to need it.

File: vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/AbstractGridMultiSelectionModel.java [hunk 2]
```diff
@@ -76,15 +76,15 @@ public AbstractGridMultiSelectionModel(Grid<T> grid) {
 
         selectionColumn
                 .setSelectAllCheckBoxVisibility(isSelectAllCheckboxVisible());
-        grid.getElement().getNode().runWhenAttached(ui -> grid.getElement()
```
Comment (tomivirkki) [inline vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/AbstractGridMultiSelectionModel.java]: Why is the deferring removed? I'm sure there was a purpose it was added there originally.

Comment (hongduc-phan) [reply vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/AbstractGridMultiSelectionModel.java]: This idea is that 
1. There is a button with addEventListener, the code  `grid.getElement().getNode().runWhenAttached(ui -> grid.getElement()...` runs immediately because the `grid.getElement()` is attached already. So the current selection mode (`grid.getSelectionModel`) is still SINGLE => the condition  `grid.getSelectionModel == this` is never true, test case: trigger change Selection Mode to MULTI by clicking button is failed. 
2. If we do not have the checking condition `grid.getSelectionModel == this`, like the code of master branch, when we change Selection mode to others like NONE/SINGLE from MULTI, there is a bug, because ` grid.getElement()
                        .insertChild(0, selectionColumn.getElement());` runs after remove method.

So my old idea is to...remove `runWhenAttached ` and the grid can insert the column checkbox.

But I changed code with a new fixing version.

File: vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/AbstractGridMultiSelectionModel.java (hunk 1 above)
Comment (hongduc-phan) [reply vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/AbstractGridMultiSelectionModel.java]: changed code with a new fixing version.

File: vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/AbstractGridMultiSelectionModel.java [hunk 3]
```diff
@@ -42,15 +42,15 @@
 import com.vaadin.flow.data.selection.SelectionEvent;
 import com.vaadin.flow.data.selection.SelectionListener;
 import com.vaadin.flow.dom.Element;
+import com.vaadin.flow.server.Command;
```
[the remark this answers is not in the document]
Comment (tulioag) [reply vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/AbstractGridMultiSelectionModel.java]: Please do that.

File: vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/AbstractGridMultiSelectionModel.java [hunk 4]
```diff
@@ -194,14 +207,14 @@ public void updateSelection(Set<T> addedItems, Set<T> removedItems) {
     @Override
     public boolean isSelected(T item) {
         return getSelectedItems().stream().anyMatch(selectedItem -> Objects
-            .equals(getItemId(selectedItem), getItemId(item)));
+                .equals(getItemId(selectedItem), getItemId(item)));
```
Comment (tulioag) [inline vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/AbstractGridMultiSelectionModel.java]: You need to fix your code formatting settings.  Import the code settings in https://github.com/vaadin/flow/tree/master/eclipse then reformat this file.

Comment (shadikhani) [reply vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/AbstractGridMultiSelectionModel.java]: You can set automatic formatting rules so your editor will automatically add spaces etc. where they should be. Here are the instructions: 
https://docs.google.com/document/d/1_1S2VILmr8y5B3hD_fMbb9SwpkFKZjnAmaW2oE4_uYo/edit#heading=h.c19b3mugjara

Comment (hongduc-phan) [reply vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/AbstractGridMultiSelectionModel.java]: fixed it back.

File: vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/AbstractGridMultiSelectionModel.java [hunk 5]
```diff
@@ -76,8 +75,22 @@ public AbstractGridMultiSelectionModel(Grid<T> grid) {
 
         selectionColumn
                 .setSelectAllCheckBoxVisibility(isSelectAllCheckboxVisible());
-        grid.getElement().getNode().runWhenAttached(ui -> grid.getElement()
-                .insertChild(0, selectionColumn.getElement()));
+
+        if (grid.getElement().getNode().isAttached()) {
+            grid.getElement().getNode().runWhenAttached(ui ->
```
[the remark this answers is not in the document]
Comment (hongduc-phan) [reply vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/AbstractGridMultiSelectionModel.java]: removed it!

[the remark this answers is not in the document]
Comment (hongduc-phan) [reply vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/AbstractGridMultiSelectionModel.java]: done.

**Comment by hongduc-phan** [comment]
> Please remove `vaadin-grid-flow-integration-tests/webpack.generated.js`
> 
> Looks like you removed it once but then committed again.

I removed it.

File: vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/grid/it/GridMultiSelectionColumnPage.java [hunk 6]
```diff
@@ -117,4 +119,30 @@ private void setUp(Grid<String> grid) {
         grid.addSelectionListener(event -> message.setText(
                 "Selected item count: " + event.getAllSelectedItems().size()));
     }
+
+    private void createBasicGridFromSingleToMulti_beforeAttached() {
```
[the remark this answers is not in the document]
Comment (tulioag) [reply vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/grid/it/GridMultiSelectionColumnPage.java]: Please follow the java conventions for naming methods. Only on integration tests they can have these long names separated by _ .

File: vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridItemRefreshPageIT.java [hunk 7]
```diff
@@ -71,7 +71,7 @@ public void dataProviderRefreshItem_gridDoesntTouchOtherItemsOnDOM() {
         GridTHTDElement firstRowCell = grid.getCell(0, 1);
 
         Assert.assertEquals("Invalid cell content", "0", firstRowCell.getText());
-        Assert.assertEquals("Invalid cell content", "5", grid.getCell(5,1).getText());
+        Assert.assertEquals("Invalid cell con0tent", "5", grid.getCell(5,1).getText());
```
Comment (tulioag) [inline vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridItemRefreshPageIT.java]: Why was this changed?

File: vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridMultiSelectionColumnPageIT.java [hunk 8]
```diff
@@ -64,6 +64,30 @@ public void selectAllCheckbox() {
                 "Selected item count: "
                         + (GridMultiSelectionColumnPage.ITEM_COUNT - 1),
                 message.getText());
+
+        // Test switch selection mode from single to multi mode before adding the grid to DOM
+        WebElement gridSelectionMode = findElement(By.id("in-testing-multi-selection-mode-grid"));
```
Comment (tulioag) [inline vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridMultiSelectionColumnPageIT.java]: This test is already big enough. This code should be in a new test method.

File: vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridItemRefreshPageIT.java (hunk 7 above)
Comment (hongduc-phan) [reply vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridItemRefreshPageIT.java]: I do not know why it added 0 within content word, I will correct it back

File: vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridMultiSelectionColumnPageIT.java (hunk 8 above)
Comment (hongduc-phan) [reply vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridMultiSelectionColumnPageIT.java]: done

File: vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridItemRefreshPageIT.java (hunk 7 above)
Comment (hongduc-phan) [reply vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridItemRefreshPageIT.java]: done

File: vaadin-grid-flow-integration-tests/webpack.generated.js [hunk 9]
```diff
@@ -0,0 +1,181 @@
+/**
```
Comment (tulioag) [inline vaadin-grid-flow-integration-tests/webpack.generated.js]: Please do not commit this file.

Comment (hongduc-phan) [reply vaadin-grid-flow-integration-tests/webpack.generated.js]: done! sorry, I forgot to delete it before pushing it

File: vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridHeaderFooterRowIT.java [hunk 10]
```diff
@@ -257,6 +259,15 @@ public void addHeaderRow_setMultiselect_disableSelection() {
                 "0", headerCells.get(0).getText());
     }
 
+    @Test
+    public void multiSelectionToSingleselect() {
+        clickButton("set-singleselect");
+        Assert.assertEquals(findElements(By.tagName("vaadin-checkbox")).stream().count(), 0);
```
Comment (tulioag) [inline vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridHeaderFooterRowIT.java]: 1. This is the method definition of `Assert.assertEquals`
```java
static public void assertEquals(long expected, long actual)
```
So the expected result should be the first parameter, and the actual result the second.

2. `findElements` returns a list, so you can use `size()` instead of `stream().count()`

File: vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridHeaderFooterRowIT.java [hunk 11]
```diff
@@ -233,6 +233,8 @@ private void assertHeaderTextsAreRendered() {
     public void addHeaderRow_setMultiselect_disableSelection() {
         clickButton("prepend-header");
         clickButton("set-multiselect");
+        WebElement gridVaadinCheckbox = findElement(By.id("selectAllCheckbox"));
+        Assert.assertTrue(String.valueOf(gridVaadinCheckbox.isDisplayed()), true);
```
Comment (tulioag) [inline vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridHeaderFooterRowIT.java]: `Assert.assertTrue(...,true)` is pointless, it will always pass.

File: vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridHeaderFooterRowIT.java [hunk 12]
```diff
@@ -257,6 +259,15 @@ public void addHeaderRow_setMultiselect_disableSelection() {
                 "0", headerCells.get(0).getText());
     }
 
+    @Test
+    public void multiSelectionToSingleselect() {
+        clickButton("set-singleselect");
+        Assert.assertEquals(findElements(By.tagName("vaadin-checkbox")).stream().count(), 0);
+        WebElement secondRow = findElements(By.tagName("vaadin-grid-cell-content")).get(4);
+        secondRow.click();
+        Assert.assertTrue(String.valueOf(secondRow.isSelected()),true);
```
Comment (tulioag) [inline vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridHeaderFooterRowIT.java]: `Assert.assertTrue(...,true)` is pointless, it will always pass.

File: vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridMultiSelectionColumnPageIT.java [hunk 13]
```diff
@@ -66,6 +66,33 @@ public void selectAllCheckbox() {
...
+    @Test
+    public void selectCheckboxMultiSelectionMode() {
+        open();
+
+        // Test switch selection mode from single to multi mode before adding the grid to DOM
+        WebElement gridSelectionMode = findElement(By.id("in-testing-multi-selection-mode-grid"));
+        WebElement selectAllCheckbox_selectionMode = gridSelectionMode
+                .findElement(By.id("selectAllCheckbox"));
+        WebElement message_selectionMode = findElement(By.id("selected-item-count"));
+        Assert.assertTrue(String.valueOf(selectAllCheckbox_selectionMode.isDisplayed()), true);
```
Comment (tulioag) [inline vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridMultiSelectionColumnPageIT.java]: `Assert.assertTrue(...,true)` is pointless, it will always pass.

File: vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridMultiSelectionColumnPageIT.java [hunk 14]
```diff
@@ -76,6 +103,11 @@ public void noSelectOnRowItemClick() {
                 .get().click();
         Assert.assertEquals("No selection event should be fired", "",
                 findElement(By.id("selected-item-count")).getText());
+
+        // Test switch selection mode from Multi to single mode before adding the grid to DOM
+        // By checking the vaadin-grid-cell-content
+        WebElement gridSelectionMode = findElement(By.id("in-testing-multi-selection-mode-grid-single"));
+        Assert.assertEquals(gridSelectionMode.findElements(By.tagName("vaadin-checkbox")).size(), 0);
```
Comment (tulioag) [inline vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridMultiSelectionColumnPageIT.java]: This is the method definition of `Assert.assertEquals`
```java
static public void assertEquals(long expected, long actual)
```
So the expected result should be the first parameter, and the actual result the second.

File: vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridHeaderFooterRowIT.java (hunk 10 above)
Comment (hongduc-phan) [reply vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridHeaderFooterRowIT.java]: done

File: vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridHeaderFooterRowIT.java (hunk 11 above)
Comment (hongduc-phan) [reply vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridHeaderFooterRowIT.java]: done

File: vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridHeaderFooterRowIT.java (hunk 12 above)
Comment (hongduc-phan) [reply vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridHeaderFooterRowIT.java]: done

File: vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridMultiSelectionColumnPageIT.java (hunk 13 above)
Comment (hongduc-phan) [reply vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridMultiSelectionColumnPageIT.java]: done

File: vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridMultiSelectionColumnPageIT.java (hunk 14 above)
Comment (hongduc-phan) [reply vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridMultiSelectionColumnPageIT.java]: done

File: vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridHeaderFooterRowIT.java (hunk 11 above)
Comment (tulioag) [reply vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridHeaderFooterRowIT.java]: `Assert.assertTrue(selectAllCheckbox_selectionMode.isDisplayed());` is more readable.
Better yet, if you can provide a meaningful error message:
```java
Assert.assertTrue("The select all checkbox should be displayed", selectAllCheckbox_selectionMode.isDisplayed());
```

File: vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridHeaderFooterRowIT.java (hunk 10 above)
Comment (tulioag) [reply vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridHeaderFooterRowIT.java]: Now that I see the fix, it occurs to me that this would be more readable:
```java
Assert.assertTrue("Error message", findElements(By.tagName("vaadin-checkbox")).isEmpty());
```
Replace "Error message" with a meaningful error message.

Comment (hongduc-phan) [reply vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridHeaderFooterRowIT.java]: fixed

File: vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridHeaderFooterRowIT.java (hunk 11 above)
Comment (hongduc-phan) [reply vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridHeaderFooterRowIT.java]: changed

File: vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridHeaderFooterRowIT.java [hunk 15]
```diff
@@ -257,6 +259,16 @@ public void addHeaderRow_setMultiselect_disableSelection() {
                 "0", headerCells.get(0).getText());
     }
 
+    @Test
+    public void multiSelectionToSingleselect() {
+        clickButton("set-singleselect");
+        grid.select(1);
+        Assert.assertTrue( "The first row of the grid should be selected", grid.getRow(1).isSelected());
```
Comment (tulioag) [inline vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridHeaderFooterRowIT.java]: Please remove the extra space after the `(`

File: vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridMultiSelectionColumnPageIT.java [hunk 16]
```diff
@@ -76,6 +103,11 @@ public void noSelectOnRowItemClick() {
                 .get().click();
         Assert.assertEquals("No selection event should be fired", "",
                 findElement(By.id("selected-item-count")).getText());
+
+        // Test switch selection mode from Multi to single mode before adding the grid to DOM
+        // By checking the vaadin-grid-cell-content
+        WebElement gridSelectionMode = findElement(By.id("in-testing-multi-selection-mode-grid-single"));
+        Assert.assertEquals(0, gridSelectionMode.findElements(By.tagName("vaadin-checkbox")).size());
```
Comment (tulioag) [inline vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridMultiSelectionColumnPageIT.java]: `Assert.assertTrue(gridSelectionMode.findElements(By.tagName("vaadin-checkbox")).isEmpty())` would be more readable.

File: vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridHeaderFooterRowIT.java [hunk 17]
```diff
@@ -257,6 +259,16 @@ public void addHeaderRow_setMultiselect_disableSelection() {
...
+        grid.select(2);
+        Assert.assertFalse( "The first row of the grid should be not selected", grid.getRow(1).isSelected());
```
Comment (tulioag) [inline vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridHeaderFooterRowIT.java]: Please remove the extra space after the `(`

File: vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridHeaderFooterRowIT.java [hunk 18]
```diff
@@ -257,6 +259,16 @@ public void addHeaderRow_setMultiselect_disableSelection() {
...
+        Assert.assertTrue( "The second row of the grid should be selected", grid.getRow(2).isSelected());
```
Comment (tulioag) [inline vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridHeaderFooterRowIT.java]: Please remove the extra space after the `(`

Comment (hongduc-phan) [reply vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridHeaderFooterRowIT.java]: done

File: vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridHeaderFooterRowIT.java (hunk 17 above)
Comment (hongduc-phan) [reply vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridHeaderFooterRowIT.java]: done

File: vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridMultiSelectionColumnPageIT.java (hunk 16 above)
Comment (hongduc-phan) [reply vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridMultiSelectionColumnPageIT.java]: done

File: vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridHeaderFooterRowIT.java (hunk 15 above)
Comment (hongduc-phan) [reply vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridHeaderFooterRowIT.java]: done
